[Login] Use magic link for suspicious emails

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #15442
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

## Description
This PR implements the first part of the project, updating the suspicious email flow to use magic links in the login.

In my spike initially, I suggested reusing the screen `LoginLinkRequestViewController`, but when working on this, I found out that this screen is not part of the unified login flow, so I added a new screen, and I matched its design to the Android one.

## Steps to reproduce
TC1: Confirming navigation to magic link request screen
1. Open the app.
2. Enter the address of a Jetpack connected site
3. Enter the email address shared here 37602-pb
4. Make sure to not send the magic link.

For all the following TCs, apply the following [patch](https://github.com/user-attachments/files/19496879/woocommerce-ios-01-49-19.patch)

TC2: Magic Link login flow
1. Open the app
2. Enter the address of a Jetpack connected site.
3. Enter an address email that can receive magic links
5. Tap on Send email in the next screen.


TC3: Fallback to username/password
1. Open the app
2. Enter the address of a Jetpack connected site.
3. Enter an address email that can receive magic links
4. Tap on Use username and password instead.
5. Sign in using site credentials.

## Testing information
TC1: Confirm that the app shows the Magic Link request screen.

TC2: Confirm that the email was sent, and that the following was tracked:
 ```
🔵 Tracked unified_login_step, properties: [flow: login_magic_link, step: start, source: default]
🔵 Tracked login_magic_link_request_form_viewed
🔵 Tracked unified_login_interaction, properties: [source: default, flow: login_magic_link, step: start, click: request_magic_link]
🔵 Tracked unified_login_step, properties: [source: default, step: magic_link_auto_requested, flow: login_magic_link]
```

TC3: Confirm that signing using site credentials work, and that the following was tracked after tapping on the fallback button:
```
Tracked unified_login_step, properties: [flow: login_magic_link, source: default, step: start]
🔵 Tracked login_magic_link_request_form_viewed
🔵 Tracked unified_login_interaction, properties: [click: login_with_wp_com_username_password, source: default, step: start, flow: login_magic_link]
```

## Screenshots
https://github.com/user-attachments/assets/994d6241-ce75-46be-9d5f-3893cf305256


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

hichamboushaba committed:
Add a new screen for requesting magic links

hichamboushaba committed:
Allow specifying the fallback button type

hichamboushaba committed:
Handle navigation to fallback screens

hichamboushaba committed:
Fix username/password login to wordpress.com

hichamboushaba committed:
Move logic to separate function

hichamboushaba committed:
Handle requesting magic link and navigation

hichamboushaba committed:
Align button text on magic link sent screen

hichamboushaba committed:
Add tracking

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/241">22.1</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba committed:
Add release note

hichamboushaba commented on WooCommerce/WordPressAuthenticator/Unified Auth/View Related/Login/MagicLinkRequestViewController.swift:
```diff
@@ -0,0 +1,230 @@
+
+// MARK: UI Setup
+private extension MagicLinkRequestViewController {
+    private func configureStackView() {
```
I didn't use a xib file, as I still don't feel familiar with XCode UI for them, and I didn't use SwiftUI just to make sure the content is correctly styled to match the other screens. If you have any remarks about this please share them.

selanthiraiyan commented on WooCommerce/WordPressAuthenticator/Unified Auth/View Related/Login/MagicLinkRequestViewController.swift:
```diff
@@ -0,0 +1,230 @@
+        )
+    }
+
+    private func fallbackButtonTitle() -> String {
```
```suggestion
    func fallbackButtonTitle() -> String {
```

nit: This method is already inside a private extension.

selanthiraiyan commented on WooCommerce/WordPressAuthenticator/Unified Auth/View Related/Login/MagicLinkRequestViewController.swift:
```diff
@@ -0,0 +1,230 @@
+        pinSubviewToHorizontalEdges(primaryButton)
+    }
+
+    private func createHeader() -> UIView {
```
nit: We can remote the private keyword as this method is already inside a private extension. Applies to other methods inside this extension.

selanthiraiyan commented on WooCommerce/WordPressAuthenticator/Unified Auth/View Related/Login/MagicLinkRequestViewController.swift:
```diff
@@ -0,0 +1,230 @@
+    private func createHeader() -> UIView {
+        let headerStackView = UIStackView()
+        headerStackView.spacing = 16
+        headerStackView.layoutMargins = UIEdgeInsets(top: 24, left: 16, bottom: 24, right: 16)
```
I suggest moving the spacing and padding related constants used in this view inside a Layout enum like we follow in SwiftUI views. WDYT?

selanthiraiyan reviewed (commented):
Thanks for working on this, Hicham! Great job navigating through the authenticator library code. 👏 


---

I have a the following questions about the UI before approving this PR. 


<img src="https://github.com/user-attachments/assets/61026ebf-d0ac-4867-a924-1bff343581ac" width="320"/>

1. Would the wording "store credentials" in the above screen create confusion about where to enter their store's wporg credentials or WPCOM username and password? 
2. I see that we pre-fill the email in the above screen. AFAIK, logging in using WPCOM email and password will not work for a suspicious email. Should we avoid pre-filling the email.


<img src="https://github.com/user-attachments/assets/c2445f13-f641-4037-ad15-69bdc733bf8a" width="320"/>


The title of "Use username and password instead" button isn't fully visible while using large fonts.

hichamboushaba committed:
Allow magic link fallback button to use multiple lines when needed

hichamboushaba committed:
Remove the unused extra private statements

hichamboushaba committed:
Use a specific intructions label for wpcom login

hichamboushaba committed:
Avoid prefilling the username field when signing in to WordPress.com

Because in this case the `loginFields.username` will hold the email address, which can't be used for login.

hichamboushaba committed:
Move layout dimensions to constants class

hichamboushaba committed:
Merge branch 'trunk' into issue/14878-login-suspicious-email-p1

hichamboushaba commented:
Thanks @selanthiraiyan for the great review, all the remarks should be addressed now, ready for another round.

itsmeichigo reviewed (approved):
Thank you for the updates Hicham!

hichamboushaba committed:
Merge branch 'trunk' into issue/14878-login-suspicious-email-p1
